[lldb][riscv] Fix setting breakpoint for undecoded instruction

ita-sc opened the pull request:
Hi

This patch adds an interface GetLastInstrSize to get information about the size of last tried to be decoded instruction and uses it to set software breakpoint if the memory can be decoded as instruction.

RISC-V architecture instruction format specifies the length of instruction in first bits, so we can set a breakpoint for these cases. This is needed as RISCV have a lot of extensions, that are not suppored by `EmulateInstructionRISCV`.

jasonmolenda commented on lldb/source/Plugins/Instruction/RISCV/EmulateInstructionRISCV.h:
```diff
@@ -99,6 +100,8 @@ class EmulateInstructionRISCV : public EmulateInstruction {
 private:
   /// Last decoded instruction from m_opcode
   DecodeResult m_decoded;
+  /// Last tried to be decoded instruction expected size.
```
I might call this `Last decoded instruction size`.

ita-sc replied:
Well, I tried to emphasize that even if we failed to fully decode instruction, we are still able to fill this information. I'll change to 
```
Last decoded instruction size estimate.
```

jasonmolenda commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -115,8 +148,23 @@ Status NativeProcessSoftwareSingleStep::SetupSoftwareSingleStepping(
 
-  if (!emulator_up->ReadInstruction())
-    return Status("Read instruction failed!");
+  if (!emulator_up->ReadInstruction()) {
```
Shouldn't this block now be `if (emulator_up->ReadInstruction())` now?  We're going to get the size of the last decoded instruction here.

jasonmolenda replied:
Ah wait, I see.  This method is trying to decode where the next instruction will go, with branches and jumps decoded, so we can put a breakpoint there.  And you're handling the case where we can't decode the current instruction (I now understand why you used that in your test case).  It seems harmless to call GetLastInstrSize() if the instruction that couldn't be decoded, and add the length of the instruction to pc.  We can assume the emulation engine will emulate all branching instructions.  I could imagine the RISCV emulation plugin didn't have decoding for an instruction that doesn't branch, it could fail, but we can still decode the size of that unknown instruction successfully, and assume that it does not branch.

jasonmolenda commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -115,8 +148,23 @@ Status NativeProcessSoftwareSingleStep::SetupSoftwareSingleStepping(
+    // try to get at least the size of next instruction to set breakpoint.
+    auto instrSizeOpt = emulator_up->GetLastInstrSize();
+    if (!instrSizeOpt)
+      return Status("Read instruction failed!");
```
We've defined the new GetLastInstrSize() method for the RISCV EmulateInstruction plugin, but others like AArch64 won't have that, so this will error out on them, won't it?  What we really want to express is "if `arch.GetTriple().isRISCV()` and we couldn't decode the length of the last instruction, then error out" isn't it?

ita-sc replied:
Well, in this MR this is a common code, and every other architecture now not implemented `GetLastInstrSize`, so it will fail with the same error, as it was before. For RISC-V we will generate the same error if we can not get the size of instruction (if we can not read memory or if an instruction is bigger than expected)

jasonmolenda commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -115,8 +148,23 @@ Status NativeProcessSoftwareSingleStep::SetupSoftwareSingleStepping(
+      return Status("Reading pc failed!");
+    lldb::addr_t next_pc = pc + *instrSizeOpt;
+    auto Result =
+        SetSoftwareBreakPointOnPC(arch, next_pc, /* next_flags */ 0x0, process);
```
We've decoded the length of the instruction at `pc` at this point, and them to get `next_pc`.  Then we pass `next_pc` to this method which has a hardcoded size of 4 for RISCV.  It's only a hint that is sent to lldb-server as it tries to step over the instruction.  With armv7/aarch32 we had to get arm/thumb breakpoint instructions correct because an arm breakpoint wasn't valid when the processor was in thumb mode (iirc) but RISC-V doesn't have a processor mode like that iiuc.  So maybe it's fine to have `SetSoftwareBreakPointOnPC` hardcoding 4 for the next RISCV breakpoint.

ita-sc replied:
Actually, I do not understand fully what you mean. For RISCV software breakpoint size is 4 or 2 (currently LLDB hardcodes only 4). We save memory and restore it later, so it seems we will have no problems here for different instruction sizes on memory regions where we write software breakpoint.

jasonmolenda reviewed (commented):
I read this patch while trying to keep the RVC instruction set in mind because it's an interesting case to consider.  Maybe the test file could have a compressed instruction?  The one instruction there now ends in 0b11 so I don't think it will decode that way.

ita-sc commented:
> I read this patch while trying to keep the RVC instruction set in mind because it's an interesting case to consider. Maybe the test file could have a compressed instruction? The one instruction there now ends in 0b11 so I don't think it will decode that way.

The test intentionally has an invalid instruction. I'll add one more test with compressed illegal instruction.
In current test it seems it is decoded as some 4-size instruction, as expected: https://godbolt.org/z/bMPf7faWW
```
addi	s0,sp,16
 .4byte	0xc58573
```

labath commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -94,6 +94,39 @@ static lldb::addr_t ReadFlags(NativeRegisterContext &regsiter_context) {
                                                  LLDB_INVALID_ADDRESS);
 }
 
+static int GetSoftwareWatchpointSize(const ArchSpec &arch,
```
Doesn't this return the size of the *break*point (i.e. it should be called `GetSoftwareBreakpointSize` ?

ita-sc replied:
Fixed, changed to `GetSoftwareBreakpointSize`

labath commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -94,6 +94,39 @@ static lldb::addr_t ReadFlags(NativeRegisterContext &regsiter_context) {
+  return 0;
+}
+
+static Status SetSoftwareBreakPointOnPC(const ArchSpec &arch, lldb::addr_t pc,
```
We usually spell "break point" as a single word (i.e., `Breakpoint`).

ita-sc replied:
Fixed

labath commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -94,6 +94,39 @@ static lldb::addr_t ReadFlags(NativeRegisterContext &regsiter_context) {
+  // space, ignore it and let the debugee segfault.
+  if (error.GetError() == EIO || error.GetError() == EFAULT) {
+    return Status();
+  } else if (error.Fail())
```
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

ita-sc replied:
Fixed

labath commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -115,8 +148,23 @@ Status NativeProcessSoftwareSingleStep::SetupSoftwareSingleStepping(
-    return Status("Read instruction failed!");
+  if (!emulator_up->ReadInstruction()) {
+    // try to get at least the size of next instruction to set breakpoint.
+    auto instrSizeOpt = emulator_up->GetLastInstrSize();
```
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

The rule is pretty subjective, but I think the fact you felt the need to add "opt" to the name shows the code is not completely understandable without it, and I'd say that spelling out the name is the more conventional way to express that. Also, (for better or worse) most of the lldb code (the surrounding code included) uses snake_case for variable name, so it's better to stick to that.

ita-sc replied:
Changed to `instr_size`

ita-sc committed:
[lldb][riscv] Fix setting breakpoint for undecoded instruction

Copy gdb behaviour:
For RISCV we can set breakpoint even for unknown instruction, as
RISCV encoding have information about size of instruction.

ita-sc commented on lldb/test/API/riscv/break-undecoded/TestBreakpointIlligal.py:
```diff
@@ -0,0 +1,27 @@
+from lldbsuite.test import lldbutil
+
+
+class TestBreakpointIlligal(TestBase):
```
Fixed

JDevlieghere commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -94,6 +94,39 @@ static lldb::addr_t ReadFlags(NativeRegisterContext &regsiter_context) {
+    if (next_flags & 0x20)
+      // Thumb mode
+      return 2;
+    else
```
`else` after return

ita-sc replied:
Fixed

JDevlieghere commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:
```diff
@@ -94,6 +94,39 @@ static lldb::addr_t ReadFlags(NativeRegisterContext &regsiter_context) {
+  return 0;
+}
+
+static Status SetSoftwareBreakpointOnPC(const ArchSpec &arch, lldb::addr_t pc,
```
Small nit, but in recent years we've been preferring using `llvm::Error` over `lldb::Status` where possible. The benefit of `Error` is that it must be checked and it's trivial to convert between the two. It would be nice if this function would return an `llvm::Error`.

ita-sc replied:
I would like to keep `Status` in this MR, as from my point of view, this MR will do more that it should -- it will add a new functionality and change the way we work with errors here.

Moreover, changing this will result in unnecessary casts from `Status` to `Error` (`process.SetBreakpoint`) and from `Error` to `Status`(`SetupSoftwareSingleStepping`).

Note: I've checked that we have `Status::ToError` and `Status(llvm::Error error)` ctor, but I'm not sure these functions will suffice (AFAIU `m_code` will be saved only if `m_type == ErrorType::eErrorTypePOSIX`).

JDevlieghere reviewed (commented):
A few small nits. I defer to Pavel and Jason for the functional changes.

ita-sc commented:
Gentle ping

labath reviewed (approved):
Thanks for pinging this. Sorry it dropped off my radar.

ita-sc commented:
Thanks. Could you please merge this? (I do not write access yet)
